src/index.js

```diff
@@ -0,0 +1,253 @@
+const events = require('events');
+const debug = require('debug');
+const amqplib = require('amqplib');
+const assert = require('assert');
+
+var connectorCounter = 0;
+
+/**
+ * Build Pulse ConnectionString, from options on the form:
+ * {
+ *   username:          // Pulse username
+ *   password:          // Pulse password
+ *   hostname:          // Hostname to use (defaults to pulse.mozilla.org)
+ * }
+ */
+const buildPulseConnectionString = function({username, password, hostname}) {
+  assert(username, 'options.username password is required');
+  assert(password, 'options.password is required');
+
+  // Construct connection string
+  return [
+    'amqps://',         // Ensure that we're using SSL
+    username,
+    ':',
+    password,
+    '@',
+    hostname || 'pulse.mozilla.org',
+    ':',
+    5671,                // Port for SSL
+  ].join('');
+};
+exports.buildPulseConnectionString = buildPulseConnectionString;
+
+/**
+ * An object to create connections to a pulse server.  This class will
+ * automatically handle reconnecting as necessary.
+ *
+ * AMQP is a very connection-oriented protocol.  For example, a client using
+ * non- durable queues will need to re-declare those queues on every new
+ * connection.  Similarly, a consumer must re-start consumption on every new
+ * connection.  This class emits a `newConnection` event on each new
+ * connection, and that function should re-establish any state as required for
+ * the new connection.
+ *
+ * Connections are automatically cycled periodically, regardless of any problems
+ * with the connection itself, in order to exercise the reconnection logic. When
+ * this occurs, the old connection is held open for 30 seconds to allow any pending
+ * publish operations or message consumptions to complete.
+ *
+ * Options:
+ * * connectionString
+ * * username
+ * * password
+ * * hostname
+ * * reconnectInterval (ms; default 1h)
+ * * retirementDelay (ms; default 30s)
+ */
+class PulseConnector extends events.EventEmitter {
+  constructor({username, password, hostname, connectionString, reconnectInterval, retirementDelay}) {
+    super();
+
+    if (connectionString) {
+      assert(!username, 'Can\'t use `username` along with `connectionString`');
+      assert(!password, 'Can\'t use `password` along with `connectionString`');
+      assert(!hostname, 'Can\'t use `hostname` along with `connectionString`');
+      this.connectionString = connectionString;
+    } else {
+      connectionString = buildPulseConnectionString({username, password, hostname});
+    }
+
+    this.reconnectInterval = reconnectInterval || 3600 * 1000; // XXX TODO
+    this.retirementDelay = retirementDelay || 30 * 1000;
+    this.running = false;
+    this.connections = [];
+    this.connectionCounter = 0;
+
+    this.id = ++connectorCounter;
+    this.debug = debug(`taskcluster-lib-pulse.connector-${this.id}`);
```

Comment:
I don't know what you mean by "global debug".  Most of the debug messages are just "connecting", so the identifier gives the context.  This will be important since there can be multiple connections at the same time.